Tests/CommandsTests/RunCommandTests.swift

```diff
@@ -102,6 +103,12 @@ final class RunCommandTests: CommandsTestCase {
     }
 
     func testSwiftRunSIGINT() throws {
+        // FIXME: not sure how this was supposed to work previously, since `swift-run` uses `execv` that doesn't inherit
+        // signal handlers, and the spawned process doesn't install signal handlers on its own. `async` effect on
+        // `@main` arguably makes it behave correctly?
```

Comment:
It seems this issue is not only a problem in tests but causes a real problem. I could reproduce this uninterruptable swift-run when invoking just built `swift-package-manager` through symlink named `swift-run`.

Based on my quick research, it seems like [libdispatch blocks `SIGINT` by `sigmask`](https://github.com/apple/swift-corelibs-libdispatch/blob/5529d351e304eb2312b7b3bfb073f7d8285473e0/src/init.c#L97) and [spawns a dedicated thread to catch those signals](https://github.com/apple/swift-corelibs-libdispatch/blob/5529d351e304eb2312b7b3bfb073f7d8285473e0/src/queue.c#L7003).

So the problematic code can be minimized as follows.

```swift
import Dispatch
import Foundation

DispatchQueue.main.async {
  let bin = "/usr/bin/sleep"
  let args = ["sleep", "10"]

  let cArgs = UnsafeMutablePointer<UnsafeMutablePointer<Int8>?>.allocate(capacity: args.count + 1)
  for (i, arg) in args.enumerated() {
    cArgs[i] = strdup(arg)
  }
  cArgs[args.count] = nil

  execv(bin, cArgs)
  perror("execv")
}

dispatchMain()
```

I confirmed we can unblock signals as follows, but I'm not 100% sure it's a safe way.

```swift
var mask = sigset_t()
sigemptyset(&mask)
pthread_sigmask(SIG_SETMASK, &mask, nil)
execv(bin, cArgs)
```

Reply:
IIUC Dispatch blocks delivery of `SIGINT` whenever some code is executed in `DispatchQueue.main.async`, is that correct? Isn't it a bug in Dispatch then and could be fixed on their side?

Reply:
No, libdispatch blocks `SIGINT` handler invocations during the whole program execution, but it creates a dedicated thread to listen to signals before blocking the signals in the main thread. `SIGINT` is usually handled by the dedicated thread and I guess the actual signal handling works are scheduled on some dispatch queues. So this is not a bug in libdispatch at all, but they don't provide a graceful way to reset signal mask states. Therefore I suggested resetting masks manually just before execv in SwiftPM as one of the options.